### ISSUE TITLE
Bromite crashes when opening links  (in incognito only mode)

### ISSUE THREAD

**Issue by AzureB1te** [bromite/bromite#899]
Running the latest version. I'm pretty sure chromium works fine. 

### Bromite version

Version: `88.4324.95.`
Arch: arm64
Android version: 11
Device model: HD1903

### Is this bug about the SystemWebView?
No

### Can the bug be reproduced with corresponding Chromium version?

Doesn't seem to be

### Is the bug a crash? 

Yes

### Describe the bug

When I click a link in apps, bromite opens up, but crashes while loading the page. Every time.

### Steps to reproduce the bug

Steps to reproduce the bug:
1. Go to any link,that opens in your browser. 
2. Click it
3. Bromite opens than crahses

### Expected behavior

Links actually open up. 

Same as #453

**Comment by a-anik** [comment]
It was like that some time ago and then it was fixed in #745. Now it's back again.
Actually, it doesn't crash when bromite is already running in the background.

**Comment by Disqu3-mirOir-qt** [comment]
same for me

**Comment by uazo** [comment]
I cannot reproduce this problem. can anyone send crash logs?

> Where do I find that?

Setttings/Info (last row)

**Comment by csagan5** [comment]
This is the same bug as #453

**Comment by hefty-fading-culprit-epiphany-drew** [comment]
Can't reproduce this on my device

**Comment by uazo** [comment]
> This is the same bug as #453, so it is a regression

only in A11? do you confirm it? do you know how to reproduce it?

**Comment by black-cerberon** [comment]
Bromite will not crash, if it will start before open link in some app. Crash will be only on 'cold start'.

My crashlog: https://del.dog/ubizibitom

**Comment by uazo** [comment]
```
time: 1612524432325
msg: Native crash: Trap
stacktrace: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
Build fingerprint: 'Xiaomi/markw/markw:6.0.1/MMB29M/V10.2.1.0.MBEMIXM:user/release-keys'
Revision: '0'
ABI: 'arm64'
pid: 24325, tid: 24325, name: bromite.bromite  >>> org.bromite.bromite <<<
signal 5 (SIGTRAP), code -6 (SI_TKILL), fault addr --------
    x0  00000076d5ac7688  x1  0000000000000000  x2  0000000000000000  x3  0000007700409208
    x4  0000000000000000  x5  0000000000000000  x6  716e7362646b6452  x7  0000007fc13adb50
    x8  0000000000000001  x9  0000000000000000  x10 0000007700400030  x11 00000001d66eb888
    x12 0000007700400040  x13 00000001d66eb880  x14 00000000ffffffff  x15 00000076d5ac7000
    x16 0000007784e9fb98  x17 0000007784e0ec50  x18 0000007700409168  x19 00000076d2307b00
    x20 00000076d2307b50  x21 00000076db3f6600  x22 00000076d2307ec0  x23 0000000000000001
    x24 00000076d2307b40  x25 00000000ffffffff  x26 0000000000000000  x27 00000076d2307b48
    x28 00000076c4798400  x29 0000007fc13add60
    sp  0000007fc13adcd0  lr  00000076c0c04e08  pc  00000076c0c05514

backtrace:
    #00 pc 00000000015fe514  /data/app/org.bromite.bromite-X6_OFqgDgoXc1xzOM1kmjg==/base.apk (offset 0xb64000)
```

too little information.
if i pass you a **mine** (not csagan5 build) debug version would you try it? can you then provide the tombstone?

**Comment by black-cerberon** [comment]
Well, I can try.

**Comment by uazo** [comment]
url: https://www.mediafire.com/file/bopfpunpa7xodr6/ChromePublic.apk/file

to extract tombstone
```
adb root
adb shell ls /data/tombstones/ -l
adb pull /data/tombstones/tombstone_<LASTNUMBER> /
```

**Comment by black-cerberon** [comment]
https://del.dog/afuvecybym.txt

**Comment by uazo** [comment]
Perfect, thanks! Crash is on `NOTREACHED() `

```
  std::string rph_crash_key_value;
  ChildProcessSecurityPolicyImpl* policy =
      ChildProcessSecurityPolicyImpl::GetInstance();
  for (RenderProcessHost::iterator host_iterator =
           RenderProcessHost::AllHostsIterator();
       !host_iterator.IsAtEnd(); host_iterator.Advance()) {
    RenderProcessHost* host = host_iterator.GetCurrentValue();
    if (host->GetBrowserContext() == this) {
      rph_crash_key_value +=
          "{ " + host->GetInfoForBrowserContextDestructionCrashReporting() +
          " }";
    }
  }
  if (!rph_crash_key_value.empty()) {
    NOTREACHED() << "rph_with_bc_reference : " << rph_crash_key_value;

    static auto* crash_key = base::debug::AllocateCrashKeyString(
        "rph_with_bc_reference", base::debug::CrashKeySize::Size256);
    base::debug::ScopedCrashKeyString auto_clear(crash_key,
                                                 rph_crash_key_value);
    base::debug::DumpWithoutCrashing();
  }
```

I'll check

**Comment by uazo** [comment]
@csagan5 this is the stacktrace. Some idea? Why does the profile close at startup? maybe `NOTREACHED` has nothing to do with it, I try to recompile it without.

```
Symbolizing stack using ABI: arm64
Build fingerprint: 'Xiaomi/markw/markw:6.0.1/MMB29M/V10.2.1.0.MBEMIXM:user/release-keys'
Revision: '0'
pid: 4288, tid: 4288, name: m.chrome.stable  >>> org.chromium.chrome.stable <<<
signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
[FATAL:browser_context.cc(494)] Check failed: false. rph_with_bc_reference : {  pl='{ https://roskomsvoboda.org/ }' lsn=2 }'
Abort message: '[FATAL:browser_context.cc(494)] Check failed: false. rph_with_bc_reference : {  pl='{ https://roskomsvoboda.org/ }' lsn=2 }'

Stack Trace:
  RELADDR   FUNCTION                                                                          FILE:LINE
  0000000000023b9c  abort+124                                                                         /system/lib64/libc.so
  v------>  base::debug::(anonymous namespace)::DebugBreak()                                  ../../base/debug/debugger_posix.cc:311:5
  0000000006c5e374  base::debug::BreakDebugger()                                                      ../../base/debug/debugger_posix.cc:344:3
  0000000006bb1cb4  logging::LogMessage::~LogMessage()                                                ../../base/logging.cc:882:7
  0000000006bb1fb0  logging::LogMessage::~LogMessage()                                                ../../base/logging.cc:558:27
  0000000004301f78  content::BrowserContext::~BrowserContext()                                        ../../content/browser/browser_context.cc:494:5
  0000000006914f74  Profile::~Profile()                                                               ../../chrome/browser/profiles/profile.cc:218:1
  0000000006913df0  OffTheRecordProfileImpl::~OffTheRecordProfileImpl()                               ../../chrome/browser/profiles/off_the_record_profile_impl.cc:229:1
  0000000006913e4c  OffTheRecordProfileImpl::~OffTheRecordProfileImpl()                               ../../chrome/browser/profiles/off_the_record_profile_impl.cc:190:53
  0000000006924490  std::__1::pair<Profile::OTRProfileID const, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >::~pair()  ../../buildtools/third_party/libc++/trunk/include/utility:297:29
  v------>  void std::__1::allocator_traits<std::__1::allocator<std::__1::__tree_node<std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >, void*> > >::__destroy<std::__1::pair<Profile::OTRProfileID const, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > > >(std::__1::integral_constant<bool, false>, std::__1::allocator<std::__1::__tree_node<std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >, void*> >&, std::__1::pair<Profile::OTRProfileID const, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >*)  ../../buildtools/third_party/libc++/trunk/include/memory:1787:23
  v------>  void std::__1::allocator_traits<std::__1::allocator<std::__1::__tree_node<std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >, void*> > >::destroy<std::__1::pair<Profile::OTRProfileID const, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > > >(std::__1::allocator<std::__1::__tree_node<std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >, void*> >&, std::__1::pair<Profile::OTRProfileID const, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >*)  ../../buildtools/third_party/libc++/trunk/include/memory:1619:14
  0000000006924a64  std::__1::__tree<std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >, std::__1::__map_value_compare<Profile::OTRProfileID, std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >, std::__1::less<Profile::OTRProfileID>, true>, std::__1::allocator<std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > > > >::erase(std::__1::__tree_const_iterator<std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >, std::__1::__tree_node<std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >, void*>*, long>)  ../../buildtools/third_party/libc++/trunk/include/__tree:2519:5
  00000000069249d4  unsigned long std::__1::__tree<std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >, std::__1::__map_value_compare<Profile::OTRProfileID, std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > >, std::__1::less<Profile::OTRProfileID>, true>, std::__1::allocator<std::__1::__value_type<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > > > >::__erase_unique<Profile::OTRProfileID>(Profile::OTRProfileID const&)  ../../buildtools/third_party/libc++/trunk/include/__tree:2542:5
  v------>  std::__1::map<Profile::OTRProfileID, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> >, std::__1::less<Profile::OTRProfileID>, std::__1::allocator<std::__1::pair<Profile::OTRProfileID const, std::__1::unique_ptr<Profile, std::__1::default_delete<Profile> > > > >::erase(Profile::OTRProfileID const&)  ../../buildtools/third_party/libc++/trunk/include/map:1304:25
  00000000069232c0  ProfileImpl::DestroyOffTheRecordProfile(Profile*)                                 ../../chrome/browser/profiles/profile_impl.cc:914:17
  000000000691e2e0  ProfileDestroyer::DestroyProfile()                                                ../../chrome/browser/profiles/profile_destroyer.cc:220:35
  0000000003348938  base::OnceCallback<void ()>::Run() &&                                             ../../base/callback.h:101:12
  0000000006c51558  base::OneShotTimer::RunUserTask()                                                 ../../base/timer/timer.cc:263:19
  0000000006c51a0c  void base::internal::Invoker<base::internal::BindState<void (base::internal::BaseTimerTaskInternal::*)(), base::internal::OwnedWrapper<base::internal::BaseTimerTaskInternal, std::__1::default_delete<base::internal::BaseTimerTaskInternal> > >, void ()>::RunImpl<void (base::internal::BaseTimerTaskInternal::*)(), std::__1::tuple<base::internal::OwnedWrapper<base::internal::BaseTimerTaskInternal, std::__1::default_delete<base::internal::BaseTimerTaskInternal> > >, 0ul>(void (base::internal::BaseTimerTaskInternal::*&&)(), std::__1::tuple<base::internal::OwnedWrapper<base::internal::BaseTimerTaskInternal, std::__1::default_delete<base::internal::BaseTimerTaskInternal> > >&&, std::__1::integer_sequence<unsigned long, 0ul>)  ../../base/bind_internal.h:710:12
  0000000003348938  base::OnceCallback<void ()>::Run() &&                                             ../../base/callback.h:101:12
  0000000006c086f0  base::TaskAnnotator::RunTask(char const*, base::PendingTask*)                     ../../base/task/common/task_annotator.cc:163:33
  0000000006c223d0  base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl(base::sequence_manager::LazyNow*)  ../../base/task/sequence_manager/thread_controller_with_message_pump_impl.cc:357:25
  0000000006c22128  base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork()   ../../base/task/sequence_manager/thread_controller_with_message_pump_impl.cc:268:36
  0000000006c6f92c  base::MessagePumpForUI::OnDelayedLooperCallback()                                 ../../base/message_loop/message_pump_android.cc:156:54
  0000000006c6f69c  base::(anonymous namespace)::DelayedLooperCallback(int, int, void*)               ../../base/message_loop/message_pump_android.cc:81:9
  000000000001446c  android::Looper::pollInner(int)+836                                               /system/lib64/libutils.so
  0000000000014084  android::Looper::pollOnce(int, int*, int*, void**)+60                             /system/lib64/libutils.so
  000000000012207c  android::android_os_MessageQueue_nativePollOnce(_JNIEnv*, _jobject*, long, int)+44  /system/lib64/libandroid_runtime.so
  00000000003aa0fc  offset 0x395000                                                                   /system/framework/arm64/boot-framework.oat
```

**Comment by uazo** [comment]
@black-cerberon sorry, not being in debug mode, that crash would not be detected in bromite, maybe it has nothing to do with this issue. 
Could you please try with https://www.mediafire.com/file/k89opc8g71mgokx/ChromePublic.apk/file ?

**Comment by black-cerberon** [comment]
@uazo may be adb logcat more usefull?
https://del.dog/oworepapam.log

**Comment by csagan5** [comment]
> only in A11? do you confirm it? do you know how to reproduce it?

I cannot reproduce it but I did not test on A11.

**Comment by csagan5** [comment]
> this is the stacktrace. Some idea? Why does the profile close at startup? maybe `NOTREACHED` has nothing to do with it, I try to recompile it without.

The call seems to be `ProfileDestroyer::DestroyProfile()`; that is from `ProfileDestroyer::RenderProcessHostDestroyed()`. I suggest you enable tracing, without that it is not trivial to debug this.

Best guess: on A11 the idle apps are treated differently, and this causes a poorly covered condition to trigger a bug.

**Comment by uazo** [comment]
> The call seems to be ProfileDestroyer::DestroyProfile(); that is from ProfileDestroyer::RenderProcessHostDestroyed(). I suggest you enable tracing, without that it is not trivial to debug this.

I don't know, the real problem could be the DCHECK related to the useragent patch, actually in production build, if null the next line would lead to a fault.
The question is, why only some users? And, how can I fix it without continuing to disturb the user? Now I'm trying to recompile for the emulator, maybe I can reproduce it.

**Comment by csagan5** [comment]
With an A11 image I am sure you can reproduce it on the emulator; perhaps user has specific settings for the UA customization.

**Comment by uazo** [comment]
> I am sure you can reproduce it on the emulator;

Nope, all is working

> perhaps user has specific settings for the UA customization.

I don't think, just installed it

this version is without that dcheck because I need more information.
can anyone provide log and tombstone?

arm64 https://www.mediafire.com/file/0giqdwip6xdd6ob/ChromePublic.apk/file
x86 https://www.mediafire.com/file/57i1ywcp42v9c0h/ChromePublic.apk/file

**Comment by a-anik** [comment]
I have this bug on A10, Oxygen OS 10.0.14.HD01AA to be specific. Same OnePlus7TP device though. It can be something specific to Oxygen OS or the launcher.

**Comment by a-anik** [comment]
Have reproduced it on my old LOS16. So the bug is not OxygenOS only.  This phone is not arm64, so I used an official build.
log:  https://del.dog/berenusyco.txt
tombstone: https://del.dog/idittulore

**Comment by uazo** [comment]
thanks @a-anik but I need log of https://github.com/bromite/bromite/issues/899#issuecomment-774477105 version, chromium tool for tombstones needs the same version as my workspace to work

**Comment by a-anik** [comment]
I found that it crashes only if another setting is also active: "Settings->Privacy->Close all open tabs on exit".
Logs for your apk:
logfile: https://del.dog/oloppapyqa
tombstone: https://del.dog/adeqoodire

**Comment by uazo** [comment]
perfect caught and reproducible! thanks a lot, now i try to solve. I'll let you know.

**Comment by uazo** [comment]
i think i solved it, can you check?

arm: https://www.mediafire.com/file/idhgyi871taj154/ChromePublic.apk/file
arm64: https://www.mediafire.com/file/l7dnx294cb1oh3e/ChromePublic.apk/file
x86: https://www.mediafire.com/file/osw1ccfs8sc31uk/ChromePublic.apk/file

the bug is caused by the changes for https://github.com/bromite/bromite/issues/654, which seem no longer necessary: can you also check that the bookmark button are active?

**Comment by a-anik** [comment]
The issue seems to be fixed in the provided arm version.  Bookmark button is active and works as intended. Well done, @uazo!

**Comment by kirk86** [comment]
Same issue here with version `88.0.4324149`, it only works when running in the background as @a-anik described. Hopefully can we expect the working version in the next update release?

**Comment by csagan5** [comment]
@kirk86 you need to mention at least your Android version.

> the bug is caused by the changes for #654, which seem no longer necessary

@uazo are you sure it is not due to different behaviour between A10 and A11?

**Comment by kirk86** [comment]
@csagan5 Sorry for that, my Android version is A10.

**Comment by uazo** [comment]
> are you sure it is not due to different behaviour between A10 and A11?

yes, tried with A9, A10 and A11, it always crashed

**Comment by csagan5** [comment]
Fixed in `88.0.4324.185`.
